### PR TITLE
Add expiry to phantom invoice utility functions

### LINKED ISSUE #1411: Include expiry in invoices
When creating an `Invoice` using functions in the `lightning-invoice` crate's `utils` module, the invoice expiration is not set. It should be set with the value encoded in the provided payment secret. Otherwise, BOLT 11 defines the default expiry to be 1 hour whereas `ChannelManager` will use whatever value is encoded in the payment secret.

This may be tricky because either the methods will need to take in a `ChannelManager` to verify the payment secret and extract the expiry or the expiry needs to be passed in but could potentially not match what's encoded in the payment secret.

### PR DESCRIPTION
**Pull request by dunxen**
Closes #1411

### CODE REVIEW

**Comment by dunxen** [comment]
Hmm. So I assumed we don't need to pass in a payment secret anymore. If this is wrong then I can fix up.

**Comment by TheBlueMatt** [comment]
I think lets drop the payment hash option as well.

File: lightning-invoice/src/utils.rs [hunk 1]
```diff
@@ -80,42 +77,34 @@ pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
...
 /// [`PhantomKeysManager`]: lightning::chain::keysinterface::PhantomKeysManager
 /// [`ChannelManager::get_phantom_route_hints`]: lightning::ln::channelmanager::ChannelManager::get_phantom_route_hints
-/// [`inbound_payment::create`]: lightning::ln::inbound_payment::create
-/// [`inbound_payment::create_from_hash`]: lightning::ln::inbound_payment::create_from_hash
 /// [`PhantomRouteHints::channels`]: lightning::ln::channelmanager::PhantomRouteHints::channels
 pub fn create_phantom_invoice_with_description_hash<Signer: Sign, K: Deref>(
-	amt_msat: Option<u64>, description_hash: Sha256, payment_hash: PaymentHash,
-	payment_secret: PaymentSecret, phantom_route_hints: Vec<PhantomRouteHints>,
-	keys_manager: K, network: Currency,
+	amt_msat: Option<u64>, payment_hash: Option<PaymentHash>, invoice_expiry_delta_secs: u32, description_hash: Sha256, phantom_route_hints: Vec<PhantomRouteHints>, keys_manager: K, network: Currency,
```
Comment (TheBlueMatt) [inline lightning-invoice/src/utils.rs:92]: Lets break the line at least a bit so its not so long :). Try to break around 100 chars.

File: lightning-invoice/src/utils.rs [hunk 2]
```diff
@@ -80,42 +77,34 @@ pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
 ///
 /// `description_hash` is a SHA-256 hash of the description text
 ///
-/// `payment_hash` and `payment_secret` can come from [`ChannelManager::create_inbound_payment`] or
+/// `payment_hash` can come from [`ChannelManager::create_inbound_payment`] or
 /// [`ChannelManager::create_inbound_payment_for_hash`]. These values can be retrieved from any
-/// participating node. Alternatively, [`inbound_payment::create`] or
-/// [`inbound_payment::create_from_hash`] may be used to retrieve these values without a
-/// `ChannelManager`.
+/// participating node. If `None` is provided for `payment_hash`, then one will be created.
```
Comment (TheBlueMatt) [inline lightning-invoice/src/utils.rs:82]: Can you add a note in the docs about what `invoice_expiry_delta_secs` is. One sentence should suffice, it is pretty clear, but best to always write it down imo.

**Review by TheBlueMatt — COMMENTED** [review]
Basically lg, is there a reason this is still draft?

**Comment by dunxen** [comment]
> Basically lg, is there a reason this is still draft?

Ah, just wanted to check on something. Thought I had made some mistake in my conceptual understand. Will promote it out of draft now :)

File: lightning-invoice/src/utils.rs [hunk 3]
```diff
@@ -821,6 +848,7 @@ mod test {
 		// Note that we have to "forward pending HTLCs" twice before we see the PaymentReceived as
 		// this "emulates" the payment taking two hops, providing some privacy to make phantom node
 		// payments "look real" by taking more time.
+
```
Comment (valentinewallace) [inline lightning-invoice/src/utils.rs:851]: nit: no newline

**Review by valentinewallace — COMMENTED** [review]
Looks great! I'm ACK if CI passes

File: lightning-invoice/src/utils.rs (hunk 3 above)
Comment (dunxen) [reply lightning-invoice/src/utils.rs:851]: Argh! :laughing:  Thanks :sweat_smile:

Comment (dunxen) [reply lightning-invoice/src/utils.rs]: Fixed :)

File: lightning-invoice/src/utils.rs [hunk 4]
```diff
@@ -38,29 +39,28 @@ use sync::Mutex;
 ///   may be too long for QR code scanning. To fix this, `PhantomRouteHints::channels` may be pared
 ///   down
 ///
-/// `payment_hash` and `payment_secret` can come from [`ChannelManager::create_inbound_payment`] or
+/// `payment_hash` can come from [`ChannelManager::create_inbound_payment`] or
```
Comment (TheBlueMatt) [inline lightning-invoice/src/utils.rs:42]: Oops, no, we don't want payment hash to come from there anymore - maybe "payment_hash can be specified if you have a specific need for a custom payment hash, see the difference between create_inbound_payment and create_inbound_payment_for_hash" (with links).

File: lightning-invoice/src/utils.rs [hunk 5]
```diff
@@ -128,6 +125,34 @@ fn _create_phantom_invoice<Signer: Sign, K: Deref>(
 		InvoiceDescription::Hash(hash) => InvoiceBuilder::new(network).description_hash(hash.0),
 	};
 
+	let keys = ExpandedKey::new(&keys_manager.get_inbound_payment_key_material());
```
Comment (TheBlueMatt) [inline lightning-invoice/src/utils.rs:128]: Maybe leave a comment here that if we ever see performance being too slow taking the `ExpandedKey` as a parameter is probably the first step.

File: lightning-invoice/src/utils.rs [hunk 6]
```diff
@@ -38,29 +39,30 @@ use sync::Mutex;
...
 ///   down
 ///
-/// `payment_hash` and `payment_secret` can come from [`ChannelManager::create_inbound_payment`] or
-/// [`ChannelManager::create_inbound_payment_for_hash`]. These values can be retrieved from any
-/// participating node. Alternatively, [`inbound_payment::create`] or
-/// [`inbound_payment::create_from_hash`] may be used to retrieve these values without a
-/// `ChannelManager`.
+/// `payment_hash` can be specified if you have a specific need for a custom payment hash (see the difference
+/// between [`ChannelManager::create_inbound_payment`] and [`ChannelManager::create_inbound_payment_for_hash`]).
+/// This value can be retrieved from any participating node. If `None` is provided for `payment_hash`, then one will be created.
```
Comment (TheBlueMatt) [inline lightning-invoice/src/utils.rs:44]: Its not "retrieved" any more - the user passes it in, I think we can drop that sentence.

**Review by TheBlueMatt — COMMENTED** [review]
Otherwise LGTM.

**Review by TheBlueMatt — APPROVED** [review]

**Review by valentinewallace — APPROVED** [review]

**Pull request merged**